File: src/code/priorityqueue/priorityqueue.js
"use strict";
function usePriorityQueue(compare = (a, b) => a === b ? 0 : (a > b ? 1 : -1)) {
    const data = [];
    function getLeftChildIndex(idx) {
        return (idx << 1) + 1;
    }
    function getRightChildIndex(idx) {
        return (idx << 1) + 2;
    }
    function getParentIndex(idx) {
        return (idx - 1) >> 1;
    }
    function swap(x, y) {
        const t = data[x];
        data[x] = data[y];
        data[y] = t;
    }
    function heapifyUp() {
        let curr = data.length - 1;
        let p = getParentIndex(curr);
        while (curr > 0 && compare(data[curr], data[p]) <= 0) {
            swap(curr, p);
            curr = p;
            p = getParentIndex(curr);
        }
    }
    function heapifyDown() {
        let curr = 0;
        while (getLeftChildIndex(curr) < data.length) {
            let next = getLeftChildIndex(curr);
            if (getRightChildIndex(curr) < data.length
                && compare(data[getLeftChildIndex(curr)], data[getRightChildIndex(curr)]) > 0)
                next = getRightChildIndex(curr);
            if (compare(data[curr], data[next]) <= 0)
                break;
            swap(curr, next);
            curr = next;
        }
    }
    function push(e) {
        data.push(e);
        heapifyUp();
    }
    function pop() {
        if (data.length === 0)
            return null;
        if (data.length === 1)
            return data.pop();
        const result = data[0];
        data[0] = data.pop();
        heapifyDown();
        return result;
    }
    const peek = () => data.length ? data[0] : null;
    const size = () => data.length;
    const isEmpty = () => size() === 0;
    return {
        push, pop, size, isEmpty, peek
    };
}

File: src/code/priority_queue/priority_queue.rs
struct PriorityQueue<T, F>
where
    F: Fn(&T, &T) -> i32,
{
    compare: F,
    data: Vec<T>,
}

impl<T, F> PriorityQueue<T, F>
where
    F: Fn(&T, &T) -> i32,
{
    fn new(compare: F) -> Self {
        Self {
            compare,
            data: Vec::new(),
        }
    }

    fn push(&mut self, value: T) {
        self.data.push(value);
        let mut i = self.data.len() - 1;
        while i > 0 {
            let parent = (i - 1) / 2;
            if (self.compare)(&self.data[i], &self.data[parent]) <= 0 {
                self.data.swap(parent, i);
                i = parent;
            } else {
                break;
            }
        }
    }

    fn pop(&mut self) -> Option<T> {
        if self.data.is_empty() {
            return None;
        }
        let result = self.data.swap_remove(0);
        let mut i = 0;
        while i * 2 + 1 < self.data.len() {
            let left = i * 2 + 1;
            let right = i * 2 + 2;
            let mut next = left;
            if right < self.data.len() && (self.compare)(&self.data[left], &self.data[right]) > 0 {
                next = right;
            }
            if (self.compare)(&self.data[next], &self.data[i]) < 0 {
                self.data.swap(next, i);
                i = next;
            } else {
                break;
            }
        }
        Some(result)
    }

    fn peek(&self) -> Option<&T> {
        self.data.first()
    }

    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    fn len(&self) -> usize {
        self.data.len()
    }

    fn clear(&mut self) {
        self.data.clear();
    }
}